tool/cmd/kitex/versions/dependencies.go

```diff
@@ -0,0 +1,266 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package versions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/cloudwego/kitex"
+	"github.com/cloudwego/kitex/tool/internal_pkg/log"
+)
+
+var dm *dependencyManager
+
+func init() {
+	dm = &dependencyManager{
+		dependencies: make(map[string]*Dependency),
+	}
+	RegisterDependencies([]Dependency{
+		// todo(DMwangnima): support reading from yaml or json files
+		{
+			RefPath:        "github.com/cloudwego/kitex",
+			MinimalVersion: "v0.9.0",
+		},
+	})
+}
+
+// RegisterDependencies registers the dependencies needed to be checked
+func RegisterDependencies(deps []Dependency) {
+	if err := dm.RegisterDependencies(deps); err != nil {
+		panic(err)
+	}
+}
+
+// CheckDependencies is responsible for checking the compatibility of dependencies
+// and return the checking results.
+// After getting the CheckResult res, you can check the error with `Err()`:
+//
+//	if err := res.Err(); err != nil {
+//	    switch {
+//	    case errors.Is(err, ErrDependencyNotFound):
+//	    case errors.Is(err, ErrDependencyVersionNotSemantic):
+//	    case errors.Is(err, ErrDependencyVersionNotCompatible):
+//	    }
+//
+// Then you can get the version in go.mod with `GoModVersion()` and retrieve the dependency
+// information with `Dependency()`.
+func CheckDependencies() []*CheckResult {
+	return dm.Check()
+}
+
+// DefaultCheckDependenciesAndProcess provided default processing procedure to parse
+// CheckResult and prompt users
+func DefaultCheckDependenciesAndProcess() {
+	var promptSb strings.Builder
+	var shouldExit bool
+	for _, res := range dm.Check() {
+		err := res.Err()
+		switch {
+		case err == nil:
+			continue
+		case errors.Is(err, ErrDependencyNotFound):
+			// ignore ErrDependencyNotFound
+		case errors.Is(err, ErrDependencyVersionNotSemantic):
+			promptSb.WriteString(fmt.Sprintf("# %s %s is not semantic version in go.mod\n", res.Dependency().RefPath, res.GoModVersion()))
+		case errors.Is(err, ErrDependencyVersionNotCompatible):
+			shouldExit = true
+			promptSb.WriteString(fmt.Sprintf("# Kitex Cmd Tool %s is not compatible with %s %s\n", kitex.Version, res.Dependency().RefPath, res.GoModVersion()))
+			promptSb.WriteString(fmt.Sprintf("# You can upgrade %s to the minimum compatible version %s\n", res.Dependency().RefPath, res.Dependency().MinimalVersion))
+			promptSb.WriteString(fmt.Sprintf("go install %s@%s\n", res.Dependency().RefPath, res.Dependency().MinimalVersion))
+			promptSb.WriteString(fmt.Sprintf("# Or downgrade Kitex Cmd Tool to %s\n", res.GoModVersion()))
+			promptSb.WriteString(fmt.Sprintf("go install github.com/cloudwego/kitex/tool/cmd/kitex@%s\n", res.GoModVersion()))
```

Comment:
If there're more than one dependency incompatible, this warning will be display several times?

Reply:
I suggest to combine it into one message listing all incompatible dependencies to update, or downgrade kitex to go mod version.

Reply:
Another problem is we need to inform ByteDance users to install the internal version so it's not a good idea to hard code the full path in this message.

Reply:
> Another problem is we need to inform ByteDance users to install the internal version so it's not a good idea to hard code the full path in this message.

This default behavior is for open-source users. We should reimplement another version of this function in internal Kitex Cmd Tool.

Reply:
And is `res.GoModVersion` always the version of `github.com/cloudwego/kitex`?

- What if we need to check also for thriftgo, netpoll, etc?
- For ByteDance users, we need to check for `code.byted.org/kite/kitex`

Reply:
```res.GoModVersion``` is the parsed dependency version in go.mod. Each ```CheckResult``` is relating to each registered ```Dependency```.